src/Compilers/CSharp/Portable/Symbols/PublicModel/TypeSymbol.cs

```diff
@@ -168,5 +169,14 @@ ImmutableArray<SymbolDisplayPart> ITypeSymbol.ToMinimalDisplayParts(SemanticMode
         bool ITypeSymbol.IsRefLikeType => UnderlyingTypeSymbol.IsRefLikeType;
 
         bool ITypeSymbol.IsReadOnly => UnderlyingTypeSymbol.IsReadOnly;
+
+        bool ITypeSymbol.IsRecord
+        {
+            get
+            {
+                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                return SynthesizedRecordClone.FindValidCloneMethod(UnderlyingTypeSymbol, ref useSiteDiagnostics) != null;
```

Comment:
>return SynthesizedRecordClone.FindValidCloneMethod(UnderlyingTypeSymbol, ref useSiteDiagnostics) != null; [](start = 16, length = 105)

I think the following approach should eliminate the infinite loop here:
1) Add ```internal abstract bool IsRecord {get;}``` to ```Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol```
2) Return result of the added property from this method.
3) This implementation goes to the override of PENamedTypeSymbol.IsRecord
4) Override new property for array, pointer, function pointer and ErrorTypeSymbol, these should return false
5) Substituted and retargeting named type symbols should return value from underlying type
6) SourceMemberContainerTypeSymbol, already defines IsRecord property. That one should be adjusted to override the new abstract property. The implementation should be good as is.
7) Override new property for other leaf (nothing derives from them) types deriving from TypeSymbol. I think the property should return false for those. I'll confirm that once the change is ready for another review pass.